Combobox.js

```diff
@@ -641,65 +665,145 @@ define([
+					items = this.value;
+				} // else empty array. No pre-set values.
+				n = items.length;
+				if (n > 1) {
+					this.inputNode.value = this.multipleChoiceMsg;
+				} else if (n === 1) {
+					this.inputNode.value = this.displayedValue !== "" ? this.displayedValue : items[0];
+				} else {
+					this.inputNode.value = this.multipleChoiceNoSelectionMsg;
+				}
+			}
+
 		},
 
 		/**
 		 * Filters the embedded List to only show the items matching `filterTxt`.
 		 * If `autoFilter` is `true` and `selectionMode` is `"single"`, the method
 		 * is called automatically while the user types into the editable input
 		 * element, with `filterTxt` being the currently entered text.
-		 * The default implementation uses `dstore/Filter.match()`.
-		 * The matching is performed against the `list.labelAttr` attribute of
-		 * the data store items.
-		 * The method can be overridden for implementing other filtering strategies.
 		 * @protected
 		 */
-		filter: function (filterTxt) {
-			if (this.filterMode === "startsWith") {
+		filter: function (inputText) {
+			// Escape special chars in search string, see
+		    // http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex.
+		    var filterTxt = inputText.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+		    if (this.filterMode === "startsWith") {
 				filterTxt = "^" + filterTxt;
 			} else if (this.filterMode === "is") {
 				filterTxt = "^" + filterTxt + "$";
 			} // nothing to add for "contains"
 
 			var rexExp = new RegExp(filterTxt, this.ignoreCase ? "i" : "");
-			this.list.query = (new Filter()).match(this.list.labelAttr, rexExp);
+			this.list.query = this.getQuery({
+				rexExp: rexExp,
+				inputText: inputText
+			});
+			if (this.source) {
+				this.list.source = this.source;
+			}
+
+			// Open popup in order to show the list's content (loading panel, no-items element or items).
+			this.openDropDown(true);
+		},
+
+		/**
+		 * Sets the new list's query.
+		 * This method can be overridden when using other store types.
+		 * The default implementation uses `dstore/Filter.match()`.
+		 * The matching is performed against the `list.labelAttr` or `list.labelFunc` attributes of
+		 * the data store items.
+		 * The method can be overridden for implementing other filtering strategies.
+		 * @protected
+		 * @returns {Object} New query to set to the list.
+		 */
+		getQuery: function (args) {
+			return (new Filter()).match(this.list.labelAttr || this.list.labelFunc, args.rexExp);
+		},
+
+		/**
+		 * Sets the initial list's query,
+		 * It can be one of a `function`, an `Object` or `null`.
+		 * If it is a function, then it's invoked and the list's query will get the return value assigned.
+		 * If it is an Object, it's assigned to the list's query directly.
+		 * If it is null, the list's query won't be set to its initial value.
+		 * It can be overridden depending of store used and the strategy to apply.
+		 */
+		getInitialQuery: {},
```

Comment:
I think we talked about this before, but I don't remember what "initial" means here.  Is it the query used when there's no text in the `<input>`?